python/tests/test_combinatorics.py

```diff
@@ -0,0 +1,668 @@
+"""
+Module for ranking/unranking labeled full non-isomorphic binary trees
+"""
+import heapq
+import math
+import random
+import unittest
+from collections import Counter
+from itertools import combinations
+
+import msprime
+
+import tskit
+import tskit.combinatorics as comb
+
+
+def L_tree(t):
+    if isinstance(t, Leaf):
+        return 1
+    n = t.num_leaves()
+    k = t.left.num_leaves()
+    total = comb.nCr(n, k) * L_tree(t.left) * L_tree(t.right)
+
+    if t.is_symmetrical():
+        assert total % 2 == 0
+        return total // 2
+    return total
+
+
+class Leaf:
+    def __init__(self, val):
+        self.val = val
+
+    def shape_rank(self):
+        return 0
+
+    def label_rank(self):
+        return 0
+
+    def label_unrank(self, rank, labels=None):
+        if labels is None:
+            labels = [0]
+
+        assert rank == 0
+        assert len(labels) == 1
+        self.val = labels[0]
+
+    def __str__(self):
+        return str(self.val)
+
+    def num_leaves(self):
+        return 1
+
+    def leaves(self):
+        return [self.val]
+
+    def height(self):
+        return 0
+
+    def shape_equal(self, other):
+        return isinstance(other, Leaf)
+
+    def equal(self, other):
+        return isinstance(other, Leaf) and other.val == self.val
+
+    def is_canonical(self):
+        return True
+
+    def canonical(self):
+        return self
+
+    @staticmethod
+    def from_tsk_tree(tree):
+        return Leaf(tree.root)
+
+    def to_tsk_tree(self):
+        tables = tskit.TableCollection(1)
+        tables.nodes.add_row(flags=tskit.NODE_IS_SAMPLE)
+        return tables.tree_sequence().first()
+
+
+class Tree:
```

Comment:
How about this alternative:
```python
class TreeNode:
    def __init__(self, children=list(), label=None):
         self.label = label
         self.children = children
         # Presumably we're building these from the leaves up? But maybe not, and that's
         # why you have a cached num_leaves value?
         self._num_leaves = None 
         self._cached_rank = None

def is_leaf(self):
      return len(self.children) == 0

def num_leaves(self):
    if self._num_leaves is None:
         if self.is_leaf():
             self._num_leaves = 1
         else:
             self._num_leaves = sum(child.num_leaves() for child in self.children)
    return self._num_leaves
```

Some rationale:

1. It seems inelegant to consider leaves and trees to be different types of things. They're really the same thing, except leaves have no children and the label is defined.
2. Using left and right is a little restrictive (maybe we'll want to define this on non-binary trees at some point). If we do keep it as left and right, I'd suggest ``left_child`` and ``right_child``. Code that uses lists of subtrees is much more elegant than left and right IMO though.
3. I was considering using a set for the children, but I think it's better to have a list in which canonically the children are ordered by the number of leaves they subtend (either in increasing or decreasing order).